Add some extra inspectors for transaction witnesses

stevenroose opened the pull request.

stevenroose committed:
witness: Improve Witness::tapscript

stevenroose committed:
witness: Add Witness::taproot_control_block

stevenroose committed:
witness: Add Witness::taproot_annex

apoelstra reviewed (dismissed):
ACK 722c3f5cd2138435c8358615a7eada5d94971b6b

Kixunil commented on bitcoin/src/blockdata/witness.rs:
```diff
@@ -384,24 +384,19 @@ impl Witness {
+            // If there are at least two witness elements, and the first byte of
+            // the last element is 0x50, this last element is called annex a
+            // and is removed from the witness stack.
+            if self.len() >= 3 && last.first() == Some(&TAPROOT_ANNEX_PREFIX) {
```
Why does it compare to 3 if it says 2? And why didn't tests trigger? A job for mutagen?

apoelstra replied:
I think the comment is just confusing. Basically, if there is only one element then it's a signature. If there are 2 or more, then the final element is a control block, unless the final element starts with `0x50`, in which case it's an annex and the second-to-last element is the control block.

Maybe the logic can be rearranged to make this clearer.

stevenroose replied:
The comment was taken from the other method that does the same thing. So yeah there can only be an annex if there are at least 3 elements (optional script witness elements (can be 0), witness script, control block and then potentially annex). I don't know how it works exactly that the control block can't start with 0x50 but I guess Sipa thought about that ;) The first byte of the control block is a flag byte anyway IIRC, so I think the 0x50 is just not a valid value for that byte.

Kixunil commented on bitcoin/src/blockdata/witness.rs:
```diff
@@ -398,6 +398,29 @@ impl Witness {
+            // If there are at least two witness elements, and the first byte of
+            // the last element is 0x50, this last element is called annex a
+            // and is removed from the witness stack.
+            if self.len() >= 3 && last.first() == Some(&TAPROOT_ANNEX_PREFIX) {
```
Again, 3 looks wrong

Kixunil reviewed (dismissed):
Disclaimer: I may be brain-dead because I don't feel great. I tried to figure out if I'm missing something but I can't see it. So this looks buggy but maybe isn't. I request that this is addressed by explaining why it's correct or fixing.

Super short commit messages and zero PR description don't help either.

tcharding commented:
> Super short commit messages and zero PR description don't help either.

Yes, please do add these. Its not super fair that I hassle ever new dev that turns up to write such things and then review PRs from you that don't have them. It also flat out makes it harder to review.

stevenroose commented:
Instead of commenting inline, let me comment here.

The 2 and 3 are maybe confusing, but they come from the difference in p2wsh and p2wpkh. 

For p2wpkh, there are 1 or 2 elements: either just signature or signature and annex.

For p2wsh, there are 2 or more elements: witness script and control block, preceded by all script witness elements and optionally succeeded by annex.

So to check for annex itself, we use 2 (to capture p2wpkh and p2wsh) but the getters for witness script and control block check for 3 because for p2wsh there can only be an annex if there's 3 elements.

stevenroose commented:
So the comment might be confusing, but it's still valid. "If there are at least 2 elements, the last one can be annex". But we also have an additional check "there need to be at least 2 non-annex elements for it to be a valid p2wsh witness". So I guess this check just combines these two conditions.

stevenroose commented:
Eh sorry, I used the term witness script instead of tapscript.

stevenroose commented:
I saw that we have a `ControlBlock` type now, so I changed the PR to use that type.

Kixunil commented on bitcoin/src/blockdata/witness.rs:
```diff
@@ -398,6 +398,40 @@ impl Witness {
+                None
+            }
+        })
+    }
```
I think we could extract this into `fn control_block_index` and then use that to index both control block and tapscript. (or `tapscript_index`, I don't care). Wouldn't do that with annex of course.

stevenroose replied:
I don't think this makes sense.

Kixunil replied:
Why not? It de-duplicates the logic which is the same. The number is just shifted by one.

Kixunil commented on bitcoin/src/blockdata/witness.rs:
```diff
@@ -691,6 +725,27 @@ mod test {
+        let witness_serialized_annex: Vec<u8> = serialize(&witness_vec_annex);
+
+        let witness = deserialize::<Witness>(&witness_serialized[..]).unwrap();
+        let witness_annex = deserialize::<Witness>(&witness_serialized_annex[..]).unwrap();
```
I don't understand the purpose of these `serialize` and `deserialize` calls. Why not `Witness::from_slice`?

stevenroose replied:
I don't know, I guess it 's just more symmetrical with the serialize call above.

stevenroose replied:
Oh, I didn't know that method existed. Fixed the tests. Thanks.

Kixunil reviewed (commented):
So this at least doesn't look broken thus I'm no longer blocking this. But I'm not giving it an ACK either because the first commit is very confusing with effectively empty commit message. It took me a long time to understand what's going on without it.

Kixunil commented:
Also if you have time, I'd love to see more tests for pathological cases.

stevenroose commented on bitcoin/src/blockdata/witness.rs:
```diff
@@ -384,24 +384,19 @@ impl Witness {
+            if self.len() >= 3 && last.first() == Some(&TAPROOT_ANNEX_PREFIX) {
+                self.nth(self.len() - 3).map(Script::from_bytes)
+            } else if self.len() >= 2 {
+                self.nth(self.len() - 2).map(Script::from_bytes)
```
sure did this

stevenroose commented:
I added some more tests and you were right that in one invalid witness case, you could still get a cb or tapscript return value. 

I fixed them, but still. These methods are not a way to validate witnesses. Just to get data from an assumed valid witness. We also don't parse if the tapscript is valid script or if the control block is an actual control block.

I think it makes sense for a library like this to make that assumptions. If anyone ever implements validation in Rust, it should be contained inside an explicit validation check library and not scattered in random getters. I was bumping into this as well when doing the script executor on Rust. Like for some things the library checks validity for others not, I think that's ok for what rust-bitcoin aims to be, but you have to be aware when you're actually trying to do validation :)

stevenroose commented:
I added some more tests and you were right that in one invalid witness case, you could still get a cb or tapscript return value. 

I fixed them, but still. These methods are not a way to validate witnesses. Just to get data from an assumed valid witness. We also don't parse if the tapscript is valid script or if the control block is an actual control block.

I think it makes sense for a library like this to make that assumptions. If anyone ever implements validation in Rust, it should be contained inside an explicit validation check library and not scattered in random getters. I was bumping into this as well when doing the script executor on Rust. Like for some things the library checks validity for others not, I think that's ok for what rust-bitcoin aims to be, but you have to be aware when you're actually trying to do validation :)

stevenroose commented:
Anyway, fixed that edge case, added tests for it and did the unwrap of the lengths.

Kixunil commented:
Yeah, it's not full validation but I think rejecting an obviously wrong input is nicer than getting a confusing error from `libbitcoinconsensus`.

apoelstra reviewed (approved):
ACK 472fe966fd0df06f60609bc784cafec49de17f51

tcharding commented on bitcoin/src/blockdata/witness.rs:
```diff
@@ -384,24 +384,72 @@ impl Witness {
+    /// This does not guarantee that this represents a P2TR [`Witness`]. See
+    /// [Script::is_p2tr](crate::blockdata::script::Script::is_p2tr) to
+    /// check whether this is actually a Taproot witness.
+    pub fn taproot_annex(&self) -> Option<&[u8]> {
```
We have an `Annex` type (in `sighash` for some reason), is there any reason why we don't use it here?

stevenroose replied:
That's kind of an empty type, right? It's clear that it's used to pass as an argument because it lives in sighash, and annexes have nothing to do with sighashes.

stevenroose replied:
The type basically just does a small invariant check that could be done inside the method that passes the annex. But yeah, I've always thought that the errors in the sighash module are actually quite annoying. I've been using it quite a lot recently and I think it merits some eyeballs.

Kixunil replied:
Looking at `Annex` it should've been unsized. But also I don't see a point in carrying around that extra byte. At best pushing it into `Witness` might be easier but really, I think there should be a trait for these things.

Anyway, I think semantically marking the bytes as `Annex` is a good idea not just for avoiding accidents but also allowing us to add methods to it once it gets used. So ACK changing the return type here.

tcharding commented:
Is this live still? Will close in two weeks unless requested otherwise.

(We are planning on [migrating away from GitHub](https://github.com/rust-bitcoin/rust-bitcoin/discussions/5714) so it is an opportune moment to clean up some stale PRs/issues.)

Kixunil commented:
This was already done more systematically.
